dashboard/modules/job/sdk.py

```diff
@@ -164,16 +193,23 @@ def submit_job(
         runtime_env = RuntimeEnv(**runtime_env).to_dict()
 
         submission_id = submission_id or job_id
-
         req = JobSubmitRequest(
             entrypoint=entrypoint,
             submission_id=submission_id,
             runtime_env=runtime_env,
             metadata=metadata,
+            num_cpus=num_cpus,
+            num_gpus=num_gpus,
+            resources=resources,
         )
 
+        json_data = dataclasses.asdict(req)
+        # Remove keys with None values so that new clients with new optional fields
+        # are still compatible with older servers.
+        json_data = {k: v for k, v in json_data.items() if v is not None}
```

Comment:
@simon-mo Can you review this and say if it's a reasonable approach to ensure forward compatibility? It feels a bit too "manual" for what seems like a very generic problem, so I wonder if there's a better way I'm missing.

WIthout this, old servers populate the json into a `JobSubmitRequest` object to validate it, so it fails when it encounters the unrecognized key `num_cpus`.

We could do this key-removing logic on the server side instead, but then we'd only have forward compatibility starting with servers running Ray 2.2+.  With the current approach, it works with older servers right off the bat.

Reply:
cc @edoakes also

Reply:
I wish there was an option to pass to `dataclasses` for this but that doesn't appear to be the case. This approach seems ok to me. Could you please add a unit test for the behavior to ensure it isn't broken in the future?

I also see no reason not to do it both on the client and server to be extra defensive. Maybe you can kill two birds with one stone and add a utility that's used both on client & server, then add a unit test for it.

Reply:
Makes sense, I'll add it to the server side as well and add a unit test